spec/unit/util/run_mode_spec.rb

```diff
@@ -216,6 +234,28 @@ def self.gte_ruby_2_4
       end
     end
 
+    describe "#public_dir" do
+      it "has publicdir ending in PuppetLabs/puppet/public when run as root" do
+        as_root { expect(@run_mode.public_dir).to eq(File.expand_path(File.join(Dir::COMMON_APPDATA, "PuppetLabs", "puppet", "public"))) }
+      end
+
+      it "has publicdir in ~/.puppetlabs/opt/puppet/public when run as non-root" do
+        as_non_root { expect(@run_mode.public_dir).to eq(File.expand_path("~/.puppetlabs/opt/puppet/public")) }
+      end
+
+      it "fails when asking for the public_dir as non-root and there is no %HOME%, %HOMEDRIVE%, and %USERPROFILE%", :unless => gte_ruby_2_4 do
```

Comment:
Why is `:unless => gte_ruby_2_4` needed?

Reply:
It was added in f4f5ccbbec53bd22d5b142cbe6f1615223cef418. I assume we're calling `without_env('HOME')` to unset the `HOME` variable from within ruby but rubies 2.4 and greater can still expand `~`.

With ruby 2.3:

```
irb(main):001:0> ENV['HOME']= nil
=> nil
irb(main):002:0> File.expand_path('~/tmp')
ArgumentError: couldn't find HOME environment -- expanding `~'
	from (irb):2:in `expand_path'
	from (irb):2
	from /usr/bin/irb:11:in `<main>'
```

With ruby 2.6:

```
2.6.3 :001 > ENV['HOME'] = nil
2.6.3 :002 > File.expand_path('~/tmp')
 => "/home/gabi/tmp" 
```

However, since this is on main and we no longer support rubies older than 2.5 this test will never run.

Reply:
Relevant ruby commit: https://github.com/ruby/ruby/commit/6b88dd2698760ef2e2306e3e79a0ed9fe159a12c

Ruby is always one step ahead of us 😆